src/main/java/com/avispl/symphony/dal/avdenvices/encoderdecoder/haivision/kraken/HaivisionKrakenCommunicator.java

```diff
@@ -0,0 +1,802 @@
+							pingResultTotal += pingResult;
+							if (this.logger.isTraceEnabled()) {
+								this.logger.trace(String.format("PING OK: Attempt #%s to connect to %s on port %s succeeded in %s ms", i + 1, host, this.getPort(), pingResult));
+							}
+						} else {
+							if (this.logger.isDebugEnabled()) {
+								this.logger.debug(String.format("PING DISCONNECTED: Connection to %s did not succeed within the timeout period of %sms", host, this.getPingTimeout()));
+							}
+							return this.getPingTimeout();
+						}
+					} catch (SocketTimeoutException | ConnectException tex) {
+						throw new RuntimeException("Socket connection timed out", tex);
+					} catch (UnknownHostException ex) {
+						throw new UnknownHostException(String.format("Connection timed out, UNKNOWN host %s", host));
+					} catch (Exception e) {
+						if (this.logger.isWarnEnabled()) {
+							this.logger.warn(String.format("PING TIMEOUT: Connection to %s did not succeed, UNKNOWN ERROR %s: ", host, e.getMessage()));
+						}
+						return this.getPingTimeout();
+					}
+				}
+				return Math.max(1, Math.toIntExact(pingResultTotal / this.getPingAttempts()));
+			} else {
+				throw new IllegalStateException("Cannot use device class without calling init() first");
+			}
+		} else {
+			throw new IllegalArgumentException("Unknown PING Mode: " + pingMode);
+		}
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	protected void authenticate() throws Exception {
+
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	protected void internalInit() throws Exception {
+		if (logger.isDebugEnabled()) {
+			logger.debug("Internal init is called.");
+		}
+		super.internalInit();
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	protected void internalDestroy() {
+		if (StringUtils.isNotNullOrEmpty(this.authenticationCookie)) {
+			deleteCookieSession();
+		}
+		if (logger.isDebugEnabled()) {
+			logger.debug("Internal destroy is called.");
+		}
+		localExtendedStatistics = null;
+		cacheValue.clear();
+		super.internalDestroy();
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	protected HttpHeaders putExtraRequestHeaders(HttpMethod httpMethod, String uri, HttpHeaders headers) throws Exception {
+		if (StringUtils.isNotNullOrEmpty(this.authenticationCookie)) {
+			headers.add(HaivisionConstant.COOKIE, "DisplayUnsavedWarning=true; Path=/; Secure;");
+			headers.add(HaivisionConstant.COOKIE, String.format("id=%s; Path=/; Secure; HttpOnly;", this.authenticationCookie));
+		}
+		return super.putExtraRequestHeaders(httpMethod, uri, headers);
+	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public List<Statistics> getMultipleStatistics() throws Exception {
+		reentrantLock.lock();
+		try {
+			Map<String, String> stats = new HashMap<>();
+			ExtendedStatistics extendedStatistics = new ExtendedStatistics();
+			if (!isEmergencyDelivery) {
+				if (!isValidCookie()) {
+					throw new FailedLoginException(String.format("Failed to login to device: %s.", this.authenticationCookie));
```

Comment:
For safety concern, I think it's better to change this message to something more generic, without adding current cookie information to it. It's not that critical, since the cookie is invalid in this case, but probably something like "...: Invalid authentication cookie" would work as well :)